runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java

```diff
@@ -110,18 +110,74 @@ final public synchronized ILogData nextUpTo(final long maxGlobal) {
         final ILogData entry =
                 getNextEntry(getCurrentContext(), maxGlobal);
 
-        // Process the next entry, checking if the context has changed.
-        // If the context has changed, we read again, since this entry
-        // does not contain any data, and we need to follow the new
-        // context.
-        if (processEntryForContext(entry)) {
-            return nextUpTo(maxGlobal);
+        if (entry != null) {
+            // Update the pointer.
+            updatePointer(entry);
+
+            // Process the next entry, checking if the context has changed.
+            // If the context has changed, we read again, since this entry
+            // does not contain any data, and we need to follow the new
+            // context.
+            if (processEntryForContext(entry)) {
+                return nextUpTo(maxGlobal);
+            }
         }
 
         // Return the entry.
         return entry;
     }
 
+    /** {@inheritDoc}
+     */
+    @Override
+    public final synchronized List<ILogData> remainingUpTo(long maxGlobal) {
+
+        // Pop the context if it has changed.
+        if (getCurrentContext().globalPointer >=
+                getCurrentContext().maxGlobalAddress) {
+            final T last = streamContexts.pollFirst();
+            log.trace("Completed context {}@{}, removing.",
+                    last.id, last.maxGlobalAddress);
+        }
+
+        final List<ILogData> entries = getNextEntries(getCurrentContext(), maxGlobal,
+                this::doesEntryUpdateContext);
+
+        // Nothing read, nothing to process.
+        if (entries.size() == 0) {
+            return entries;
+        }
+
+        // Check if the last entry updates the context.
+        if (doesEntryUpdateContext(entries.get(entries.size() - 1)))
+        {
+            // The entry which updates the context must be the last one, so
+            // process it
+            processEntryForContext(entries.get(entries.size() - 1));
+
+            // Remove the entry which updates the context
+            entries.remove(entries.size() - 1);
+
+            // do a read again, which will consume the inner context
+            entries.addAll(remainingUpTo(maxGlobal));
+
+            // and now read again, in case the context returned, to make
+            // sure we get up to maxGlobal.
+            entries.addAll(remainingUpTo(maxGlobal));
+
+            // Update the pointer with the last entry
+            updatePointer(entries.get(entries.size() - 1));
+
+            return entries;
+        }
+
+        // Otherwise update the pointer with the last entry
```

Comment:
make this the 'if' case, and return. Then if you didn't return, handle nested contexts.